plugins/aea-ledger-ethereum/aea_ledger_ethereum/ethereum.py

```diff
@@ -61,6 +64,142 @@
+
+    if base_fee_gwei <= 40:
+        return 2.0
+    elif base_fee_gwei <= 100:
+        return 1.6
+    elif base_fee_gwei <= 200:
+        return 1.4
+    else:
+        return 1.2
+
+
+def estimate_priority_fee(
+    web3: Web3,
+    base_fee_gwei: int,
+    block_number: int
+) -> int:
+    """Estimate priority fee from base fee."""
+
+    if base_fee_gwei < PRIORITY_FEE_ESTIMATION_TRIGGER:
+        return DEFAULT_PRIORITY_FEE
+
+    fee_history = web3.eth.fee_history(
+        FEE_HISTORY_BLOCKS,
+        block_number,
+        [FEE_HISTORY_PERCENTILE]
+    )
+
+    rewards = sorted([reward for reward in fee_history["reward"] if reward > 0])
+    if not len(rewards):
+        return None
+
+    # Calculate percentage increases from between ordered list of fees
+    percentage_increases = [((j - i) / i) * 100 for i,
+                            j in zip(rewards[:-1], rewards[1:])]
+    highest_increase = max(*percentage_increases)
+    highest_increase_index = percentage_increases.index(highest_increase)
+
+    values = rewards.copy()
+    # If we have big increase in value, we could be considering "outliers" in our estimate
+    # Skip the low elements and take a new median
+    if highest_increase > PRIORITY_FEE_INCREASE_BOUNDARY and highest_increase_index >= len(values) // 2:
+        values = values[highest_increase_index:]
+
+    return values[len(values) // 2]
+
+
+def get_gas_price_strategy_eip1559() -> Callable[[Web3, TxParams], Dict[str, Wei]]:
+    """Get the gas price strategy."""
+
+    def gas_station_gas_price_strategy_eip1559(  # pylint: disable=redefined-outer-name,unused-argument
+        web3: Web3, transaction_params: TxParams
+    ) -> Dict[str, Wei]:
+        """
+        Get gas price from Eth Gas Station api.
+
+        Visit `https://docs.ethgasstation.info/gas-price` for documentation.
+        :param web3: web3 instance
+        :param transaction_params: transaction parameters
+        :return: dictionary containing gas price strategy
+        """
+
+        latest_block = web3.eth.get_block("latest")
+        base_fee = latest_block.get("baseFeePerGas")
+        block_number = latest_block.get("number")
+        base_fee_gwei = wei_to_gwei(base_fee)
+
+        estimated_priority_fee = estimate_priority_fee(
+            web3,
+            base_fee_gwei,
+            block_number
+        )
+
+        if estimated_priority_fee is None:
+            _default_logger.warning(
+                "An error occurred while estimating priority fee, falling back")
+            return FALLBACK_ESTIMATE
+
+        max_priority_fee_per_gas = max(estimated_priority_fee, DEFAULT_PRIORITY_FEE)
+        multiplier = get_base_fee_multiplier(base_fee_gwei)
+
+        potential_max_fee = base_fee * multiplier
+        max_fee_per_gas = (
+            (potential_max_fee + max_priority_fee_per_gas)
+            if max_priority_fee_per_gas > potential_max_fee
+            else potential_max_fee
+        )
```

Comment:
what's the logic behind this?

Reply:
I copied the logic from the wallet implementation so I'm not sure about it, I'm still learning about EIP1559. I think this logic is to prevent the value of `max_priority_fee_per_gas` from getting larger than the value of `max_fee_per_gas` since `priority_fee` should always be less than `base_fee`.